lib/libcyr_cfg.c

```diff
@@ -213,12 +225,56 @@ EXPORTED void libcyrus_config_setswitch(enum cyrus_opt opt, int val)
     cyrus_options[opt].val.b = val;
 }
 
+/* This keeps a list of functions to call at an opportune time when the
+ * user is not waiting.
+ * Arguments:
+ *  dedup key: if provided, don't add this callback again if there's
+ *             already one for this key.
+ *  cb:        callback to call with rock passed
+ *  free:      function to call with rock after the callback to free it,
+               if NULL there is nothing to free.
+ *  rock:      arguments for the callback
+ */
+EXPORTED void libcyrus_delayed_action(const char *key, void (*cb)(void *),
+                                      void (*myfree)(void *), void *rock)
+{
+    struct delayed_action *action;
+    if (key) {
+        // check if we already have this event on our list
+        for (action = delayed_actions; action; action = action->next) {
+            if (!strcmpsafe(key, action->key)) {
+                myfree(rock);
+                return;
+            }
+        }
+    }
+    action = xzmalloc(sizeof(struct delayed_action));
+    action->key = xstrdupnull(key);
+    action->cb = cb;
+    action->myfree = myfree;
+    action->rock = rock;
+    action->next = delayed_actions;
+    delayed_actions = action;
+}
+
+EXPORTED void libcyrus_run_delayed(void)
+{
+    while (delayed_actions) {
+        struct delayed_action *action = delayed_actions;
+        delayed_actions = action->next;
+        action->cb(action->rock);
+        action->myfree(action->rock);
```

Comment:
myfree is allowed to be NULL, so need to check for that before calling it

Reply:
yes good point!